### PR TITLE
[wip] programs/nix-index.nix: init module

### PR DESCRIPTION
**Pull request by matthewbauer**
Add the nix-index module that will do two things:

- generate an index of all of the files in the binary cache of your channel
- add a command-not-found script for bash & zsh

Note that this will run nix-index on every nixos-rebuild. For it to work correctly you need plenty of memory & time.

### CODE REVIEW

File: nixos/modules/programs/nix-index.nix [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+  cfg = config.programs.nix-index;
+
+  # Wrap nix-locate to use cache
+  nix-locate = pkgs.runCommand "nix-locate" {
+    buildInputs = [pkgs.makeWrapper];
+  } ''
+mkdir -p $out/bin
+makeWrapper ${pkgs.nix-index}/bin/nix-locate $out/bin/nix-locate \
+  --add-flags "-d ${cache}"
+  '';
```
[the remark this answers is not in the document]
Comment (matthewbauer) [reply nixos/modules/programs/nix-index.nix]: Lol probably not

**Comment by jtojnar** [comment]
Could fish support be added as well? How does the regular command-not-found work with fish?

**Comment by Infinisil** [comment]
I think it would make more sense to create a package from this, so you can do `(import <nixpkgs> {}).nix-locate-cached`. Completion scripts can be installed in the shells respective path in $out. This would also allow non-NixOS users to use this.

Have you done some more work on this? I might try to help with this when I have time.

**Comment by matthewbauer** [comment]
> I think it would make more sense to create a package from this, so you can do (import <nixpkgs> {}).nix-locate-cached. Completion scripts can be installed in the shells respective path in $out. This would also allow non-NixOS users to use this.

I'm not sure if that would work easily. This cannot be built in Hydra because it requires network access. A module makes the most sense to me until we can convince everyone that this should be shipped automatically in the channels in place of programs.sqlite. You only want to run nix-index when ```pkgs.path``` is changed.

**Pull request closed without merging**